demo/terraform/main.tf

```diff
@@ -0,0 +1,224 @@
+  type                      = "pki"  
+  default_lease_ttl_seconds = 3600 
+  max_lease_ttl_seconds     = 311040000  # Maximum lease time (~10 years).
+  description               = "PKI secrets engine" 
+
+  depends_on = [null_resource.wait_for_vault]
+}
+
+# Create a self-signed root certificate.
+resource "vault_pki_secret_backend_root_cert" "root_cert" {
+  backend = vault_mount.pki.path  
+  type    = "internal"
+
+  common_name = "svc"  
+  ttl         = "87600h"  # Time-to-live for the certificate (~10 years).
+
+  depends_on = [vault_mount.pki]
+}
+
+# Store the root CA certificate locally.
+resource "local_file" "ca_cert" {
+  content  = vault_pki_secret_backend_root_cert.root_cert.certificate
+  filename = "${path.module}/CA_cert.crt"
+
+  depends_on = [vault_pki_secret_backend_root_cert.root_cert]
+}
+
+# Configure the URLs for the Vault PKI secrets engine.
+resource "vault_pki_secret_backend_config_urls" "pki_urls" {
+  backend = vault_mount.pki.path
+
+  issuing_certificates    = ["${var.vault_address}/v1/${vault_mount.pki.path}/ca"]
+  crl_distribution_points = ["${var.vault_address}/v1/${vault_mount.pki.path}/crl"]
+
+  depends_on = [vault_pki_secret_backend_root_cert.root_cert]
+}
+
+# Enable an intermediate PKI secrets engine.
+resource "vault_mount" "pki_int" {
+  path                      = "pki_int"
+  type                      = "pki"
+  default_lease_ttl_seconds = 3600
+  max_lease_ttl_seconds     = 311040000
+  description               = "PKI intermediate secrets engine"
+
+  depends_on = [vault_pki_secret_backend_config_urls.pki_urls]
+}
+
+# Create a Certificate Signing Request (CSR) for the intermediate CA.
+resource "vault_pki_secret_backend_intermediate_cert_request" "pki_int" {
+  backend = vault_mount.pki_int.path
+  type    = "internal"
+
+  common_name = "svc Intermediate Authority" 
+
+  depends_on = [vault_mount.pki_int]
+}
+
+# Store the CSR locally for signing.
+resource "local_file" "pki_intermediate_csr" {
+  content  = vault_pki_secret_backend_intermediate_cert_request.pki_int.csr
+  filename = "${path.module}/pki_intermediate.csr"
+
+  depends_on = [vault_pki_secret_backend_intermediate_cert_request.pki_int]
+}
+
+# Sign the intermediate CSR with the root CA to generate the intermediate certificate.
+resource "vault_pki_secret_backend_root_sign_intermediate" "pki_int_signed" {
+  backend = vault_mount.pki.path  # Root CA backend.
+  csr     = vault_pki_secret_backend_intermediate_cert_request.pki_int.csr
+
+  common_name = "svc Intermediate Authority"
+  ttl         = "43800h"  # TTL for the intermediate cert (~5 years).
+  format      = "pem_bundle"
+
+  depends_on = [
+    vault_pki_secret_backend_intermediate_cert_request.pki_int,
+    vault_mount.pki
+  ]
+}
+
+# Store the signed intermediate certificate locally.
+resource "local_file" "intermediate_cert_pem" {
+  content  = vault_pki_secret_backend_root_sign_intermediate.pki_int_signed.certificate
+  filename = "${path.module}/intermediate.cert.pem"
```

Comment:
The naming of the files is inconsistent IMO:

- `intermediate.cert.pem`
- `pki_intermediate.csr`
- `CA_cert.crt`

Reply:
We might want to add the name of these files to the `.gitignore` as well.